File: lib/data/region.dart
//  Licensed under Apache License v2.0

part of geolocation;

/// Represents a circular region
class Region {
  Region({@required this.center, @required this.radius});

  /// Center of the circular region
  final Location center;

  /// Radius measured in meters.
  final double radius;

  @override
  String toString() {
    return '{center: $center, radius: $radius}';
  }
}

class GeofenceRegion {
  final Region region;
  final String id;
  final bool notifyOnEntry;
  final bool notifyOnExit;

  GeofenceRegion(
      {@required this.region,
      @required this.id,
      this.notifyOnEntry = true,
      this.notifyOnExit = false});

  @override
  String toString() {
    return '{region: $region, id: $id, notifyOnEntry: $notifyOnEntry, notifyOnExit: $notifyOnExit}';
  }
}

enum GeofenceEventType { entered, exited }

class GeofenceEvent {
  final GeofenceEventType type;
  final GeofenceRegion geofenceRegion;

  GeofenceEvent._(this.type, this.geofenceRegion);
}

class GeofenceEventResult extends GeolocationResult {
  GeofenceEventResult._(
      bool isSuccessful, GeolocationResultError error, this.geofenceEvent)
      : super._(isSuccessful, error);

  final GeofenceEvent geofenceEvent;

  @override
  String dataToString() {
    return geofenceEvent.toString();
  }
}

File: ios/Classes/Data/Region.swift
//
//  Copyright (c) 2018 Loup Inc.
//  Licensed under Apache License v2.0
//

import Foundation
import CoreLocation

struct Region: Codable {
  let center: Location
  let radius: Double
  
  init(from region: CLCircularRegion) {
    self.center = Location(from: region.center)
    self.radius = region.radius
  }
}

struct GeofenceRegion: Codable {
  let region: Region
  let id: String
  let notifyOnEntry: Bool
  let notifyOnExit: Bool
  
  init(from region: CLCircularRegion) {
    self.region = Region(from: region)
    self.id = region.identifier
    self.notifyOnEntry = region.notifyOnEntry
    self.notifyOnExit = region.notifyOnExit
  }
  
  var clRegion: CLCircularRegion {
    let result = CLCircularRegion(center: self.region.center.coordinate2D, radius: self.region.radius, identifier: self.id)
    result.notifyOnExit = self.notifyOnExit
    result.notifyOnEntry = self.notifyOnEntry
    return result
  }
}

enum GeofenceEventType: String, Codable {
  case entered, exited
}

struct GeofenceEvent: Codable {
  let type: GeofenceEventType
  let geofenceRegion: GeofenceRegion
  
  init(region: CLCircularRegion, type: GeofenceEventType) {
    self.type = type
    self.geofenceRegion = GeofenceRegion(from: region)
  }
}
